lib/sda/processor.go

```diff
@@ -0,0 +1,125 @@
+package sda
+
+import (
+	"errors"
+	"reflect"
+
+	"github.com/dedis/cothority/lib/dbg"
+	"github.com/dedis/cothority/lib/network"
+)
+
+// ServiceProcessor allows for an easy integration of external messages
+// into the Services. You have to embed it into your Service-structer,
+// then it will offer an 'AddMessage'-method that takes a message of type
+// 	func ReceiveMsg(e *network.Entity, msg *anyMessageType)(error, *replyMsg)
+// where 'ReceiveMsg' is any name and 'anyMessageType' will be registered
+// with the network. Once 'anyMessageType' is received by the service,
+// the function 'ReceiveMsg' should return an error and any 'replyMsg' it
+// wants to send.
+type ServiceProcessor struct {
+	functions map[network.MessageTypeID]interface{}
+	Context
+}
+
+// NewServiceProcessor initializes your ServiceProcessor.
+func NewServiceProcessor(c Context) *ServiceProcessor {
+	return &ServiceProcessor{
+		functions: make(map[network.MessageTypeID]interface{}),
+		Context:   c,
+	}
+}
+
+// AddMessage puts a new message in the message-handler
+func (p *ServiceProcessor) AddMessage(f interface{}) error {
+	ft := reflect.TypeOf(f)
+	// Check we have the correct channel-type
+	if ft.Kind() != reflect.Func {
+		return errors.New("Input is not function")
+	}
+	if ft.NumIn() != 2 {
+		return errors.New("Need two arguments: *network.Entity and *struct")
+	}
+	if ft.In(0) != reflect.TypeOf(&network.Entity{}) {
+		return errors.New("First argument must be *network.Entity")
+	}
+	cr1 := ft.In(1)
+	if cr1.Kind() != reflect.Ptr {
+		return errors.New("Second argument must be a *pointer* to struct")
+	}
+	if cr1.Elem().Kind() != reflect.Struct {
+		return errors.New("Second argument must be a pointer to *struct*")
+	}
+	if ft.NumOut() != 2 {
+		return errors.New("Need 2 return values: network.ProtocolMessage and error")
+	}
+	if ft.Out(0) != reflect.TypeOf((*network.ProtocolMessage)(nil)).Elem() {
+		return errors.New("Need 2 return values: *network.ProtocolMessage* and error")
+	}
+	if ft.Out(1) != reflect.TypeOf((*error)(nil)).Elem() {
+		return errors.New("Need 2 return values: network.ProtocolMessage and *error*")
+	}
+	// Automatic registration of the message to the network library.
+	dbg.Lvl3("Registering handler", cr1.String())
+	typ := network.RegisterMessageUUID(network.RTypeToMessageTypeID(
+		cr1.Elem()),
+		cr1.Elem())
+	p.functions[typ] = f
+	return nil
+}
```

Comment:
No, this one is more elaborate ;) In fact the `Protocol`-reflection code searches for a specific message-structure (that I REALLY would like to change but never managed to), which is different here. Also do we have return values that need to be checked and used, so it's not the same code.